AB#16252 - Email notifications next

AntoineRelief opened the pull request:
# Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

## Useful links

- [Please insert link to ticket](https://dev.azure.com/WHOHQ/EMSSAFE/_workitems/edit/16252)
- [Please insert link to back-end branch if any](https://github.com/ReliefApplications/ems-backend/pull/1005)
- Please insert any useful link ( documentation you used for example )

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Improvement (refactor or addition to existing functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce.

- [ ] Test A
- [ ] Test B

## Screenshots

Please include screenshots of this change. If this issue is only back-end related, and does not involve any visual change of the platform, you can skip this part.

# Checklist:

( * == Mandatory ) 

- [ ] * I have set myself as assignee of the pull request
- [ ] * My code follows the style guidelines of this project
- [ ] * Linting does not generate new warnings
- [ ] * I have performed a self-review of my own code
- [ ] * I have put the ticket for review, adding the oort-frontend team to the list of reviewers
- [ ] * I have commented my code, particularly in hard-to-understand areas
- [ ] * I have put JSDoc comment in all required places
- [ ] * My changes generate no new warnings
- [ ] * I have included screenshots describing my changes if relevant
- [ ] * I have selected labels in the Pull Request, according to the changes with code brings
- [ ] I have made corresponding changes to the documentation ( if required )
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

# More explanation
https://www.loom.com/share/05a716d61b9744faaf51fb304c21d1e5?sid=f87cf896-582a-4f76-93ae-8ceed801b145

HasnatL-Adapptlimited commented:
Hi @AntoineRelief, @MwanPygmay, @matheus-relief 

Hope you're doing well!

These changes have been pushed:

## Data Changes
- Resource, Resources, CreatedBy nested fields, data persisting in preview pages and final email
- Operators populating correctly for resource and resources
- Field select now grouped and sorted correctly

## Style Fixes
- Dual List Implemented instead of click for field select
- Converted the landing page in to ui-tab view
- Search bar consistent with rest of app builder
- Modified the "Save and Proceed" button to Next
- Back button now uses ui-button
- Removed or use existing text from step 1 under email notification stepper page
- Added tool tips under the data set page for add and delete icons
- + Icon Moved out of the header
- Added the tooltip and text in distribution list page that reads "Type Email and hit ENTER to add
- Dropdown on email only provides email as option
- UI-checkbox used
- UI-Alert messaged moved to the top of page instead of at the bottom of distribution list page (step 4)

## To be delivered
- Distribution List Search on email notification page
- Send Seperate Email full functionality
- Createdby + lastupdatedby persiting in final email
- Replacing NG-Select with ui-select-menu on dataset select
- Replacing tabs with UI-tabs for dataset page

AntoineRelief commented:
@HasnatL-Adapptlimited 
In last commit, I reverted changes made on tooltip, as it was breaking the code

If you need them, please put them back, making sure the code can still build
also, I'm not sure about these changes and why they were done, so if we need them again, please put more explanation in a comment

HasnatL-Adapptlimited commented:
> @HasnatL-Adapptlimited In last commit, I reverted changes made on tooltip, as it was breaking the code
> 
> If you need them, please put them back, making sure the code can still build also, I'm not sure about these changes and why they were done, so if we need them again, please put more explanation in a comment

Tested on both chrome and firefox, tooltips seems to be working fine with revert, thank you

AntoineRelief commented:
@HasnatL-Adapptlimited 

FYI, I merged the last version of the 2.x.x branch in both front-office & back-office, so I could do another deployment on our UAT environment, and it's a good practice to keep the base branch up to date

please note that I didn't find any issue while doing so, but you may need to pull the branch before pushing any new commit

HasnatL-Adapptlimited commented:
> @HasnatL-Adapptlimited
> 
> FYI, I merged the last version of the 2.x.x branch in both front-office & back-office, so I could do another deployment on our UAT environment, and it's a good practice to keep the base branch up to date
> 
> please note that I didn't find any issue while doing so, but you may need to pull the branch before pushing any new commit

Hey @AntoineRelief, Yes thats fine, I saw and rebased the new fixes ontop of these changes, I will have another PR raised by the EOD with notes for the frontend and backend, thank you,

**Question:**
Would it be better if I raise and accept and merge the PR into this branch? or just raise the pr for each new bug fix?

AntoineRelief commented:
> > @HasnatL-Adapptlimited
> > FYI, I merged the last version of the 2.x.x branch in both front-office & back-office, so I could do another deployment on our UAT environment, and it's a good practice to keep the base branch up to date
> > please note that I didn't find any issue while doing so, but you may need to pull the branch before pushing any new commit
> 
> Hey @AntoineRelief, Yes thats fine, I saw and rebased the new fixes ontop of these changes, I will have another PR raised by the EOD with notes for the frontend and backend, thank you,
> 
> **Question:** Would it be better if I raise, and merge the PR into this branch? or just raise the pr for each new bug fix?

@HasnatL-Adapptlimited 

I think you can keep one PR per bug, so we can better track it

Matthis-M-ReliefApps commented:
Hi @HasnatL-Adapptlimited,
I'm still working on the code review itself, but here are a few feedbacks on the functional side that I can already give:

High Priority:

1. On preview page, always show the distribution list and recipients, to avoid bad surprises for the users (in case it was modified by another user for example).
2. When body is filled, then emptied, the validator doesn't always trigger and we can proceed with empty body.
3. It is possible to create a block from a dataset, but not set the fields, so it wouldn't show anything. There should probably be a validator there, unless I'm mistaking?
4. Per my understanding, the "Style" and "Send separate email" are features planned for the next phase. I recommend indicating that those features are disabled so the users don't think it's a bug (maybe through a tooltip or text on hovering). This also applies to "Scheduled alert" feature and others I might be missing.

Normal Priority:

1. To get a more consistent UX, reuse the breadcrumbs as they are used in the roles settings page, also with the browser's back button working.

Matthis-M-ReliefApps commented on libs/shared/src/lib/components/email/email.component.ts:
```diff
@@ -0,0 +1,682 @@
+          const filteredEmailList: string[][] =
+            this.emailService.emailNotificationNames
+              .map((notification: string) => notification.split('_clone'))
+              .filter(
+                (cloneNotificationName) =>
+                  cloneNotificationName.length > 1 &&
+                  cloneNotificationName[0] ===
+                    emailData.name.toLowerCase().split('_clone')[0]
+              );
+          if (filteredEmailList.length > 0) {
+            filteredEmailList.forEach(
+              (filteredEmail: string[], index: number) => {
+                if (
+                  filteredEmail[0] ===
+                  emailData.name.trim().toLowerCase().split('_clone')[0]
+                ) {
+                  if (filteredEmail[filteredEmail.length - 1] !== '') {
+                    maxCloneNumber =
+                      maxCloneNumber <
+                      Number(filteredEmail[filteredEmail.length - 1])
+                        ? Number(filteredEmail[filteredEmail.length - 1]) + 1
+                        : maxCloneNumber;
+                  } else if (
+                    filteredEmail[1] === '' &&
+                    maxCloneNumber === 0 &&
+                    index === filteredEmailList.length - 1
+                  ) {
+                    maxCloneNumber = 1;
+                  }
+                }
+              }
+            );
+            if (maxCloneNumber > 0) {
+              emailData.name =
+                emailData.name.split('_Clone')[0] + '_Clone' + maxCloneNumber;
+            } else {
+              emailData.name = emailData.name + '_Clone';
+            }
+          } else {
+            emailData.name = emailData.name + '_Clone';
```
I didn't double-check by running it in the app, but is it normal that you're writing both "_clone" and "_Clone"?

Matthis-M-ReliefApps commented on libs/shared/src/lib/components/email/email.component.ts:
```diff
@@ -0,0 +1,682 @@
+    };
+    this.emailService.tabs = [
+      {
+        title: `Block 1`,
+        content: `Block 1 Content`,
```
maybe translation here too

Matthis-M-ReliefApps commented on libs/shared/src/lib/components/email/email.component.ts:
```diff
@@ -0,0 +1,682 @@
+      title: this.translate.instant('common.deleteObject', {
+        name: this.translate.instant('common.email.notification.one'),
+      }),
+      content: 'Do you confirm the deletion of ' + data.name + ' ?',
```
missing translation

HasnatL-Adapptlimited commented:
> > > @HasnatL-Adapptlimited
> > > FYI, I merged the last version of the 2.x.x branch in both front-office & back-office, so I could do another deployment on our UAT environment, and it's a good practice to keep the base branch up to date
> > > please note that I didn't find any issue while doing so, but you may need to pull the branch before pushing any new commit
> > 
> > 
> > Hey @AntoineRelief, Yes thats fine, I saw and rebased the new fixes ontop of these changes, I will have another PR raised by the EOD with notes for the frontend and backend, thank you,
> > **Question:** Would it be better if I raise, and merge the PR into this branch? or just raise the pr for each new bug fix?
> 
> @HasnatL-Adapptlimited
> 
> I think you can keep one PR per bug, so we can better track it

For todays PR is it okay if we push 1 large PR with notes linking to each bug?
Then from tomorrow onwards we create atomic PR's for each bug?

Matthis-M-ReliefApps commented on libs/shared/src/lib/components/email/filter/filter.constant.ts:
What's the reason for not reusing "libs/shared/src/lib/components/filter/filter.const.ts"? Could there be cases where we want to update one but not the other?

AntoineRelief commented:
@HasnatL-Adapptlimited 

sure, that's fine for me

Matthis-M-ReliefApps reviewed (changes requested):
Hi @HasnatL-Adapptlimited,
I finished this round of review of the frontend.
Sorry for the huge amount of comments but don't worry, 90% of them are very small and easy-to-fix issues.
Please check the "High Priorities" ones first, they could in my opinion be blockers for the deployment. Let me know if any of my comments is lacking clarity.
Also, my general recommendation for the next developments would be to use the app builder's existing components as often as possible. Sometimes it takes a bit more time to find the right structure to fit them in, but it would really reduce duplication, and make maintenance easier.
In any case, well played for all the work done on this big feature, we're getting closer to production!

AntoineRelief commented:
@HasnatL-Adapptlimited 

In last commit:
- removed useless full import of kendo bootstrap. It was breaking some style, so I preferred to only import additional style for the kendo tabs
- replaced one UI element by a divider, to reuse existing UI elements

cc @morgang-adappt

AntoineRelief commented:
@HasnatL-Adapptlimited 

In last commit, use chip to indicate legacy pages ( distribution lists & templates )

cc @morgang-adappt

AntoineRelief commented:
@HasnatL-Adapptlimited 

in last commit
some UI improvements

@morgang-adappt

HasnatL-Adapptlimited commented on libs/shared/src/lib/components/email/components/dataset-filter/dataset-filter.component.ts:
We will try to reuse in phase 2

AntoineRelief commented:
@HasnatL-Adapptlimited 
in last commit:
- refactor create-notification component to remove duplicated css / duplicated code
- refactor create-notification component to better use angular forms API
- refactor create-notification component to hide email notifications type dropdown ( based on this feedback https://dev.azure.com/WHOHQ/EMSSAFE/_workitems/edit/90480 )

AntoineRelief commented:
@HasnatL-Adapptlimited 

In last commit:
- refactor schedule alert component, to avoid counter counter intuitive logic ( schedule_alert set to true to represent a case where it is false )
